### PR TITLE
Senza patch & respawn-instances

### PR DESCRIPTION
**Pull request by hjacobs**
Implementing #131: senza patch, senza scale and senza respawn-instances

No unit tests (yet).

### CODE REVIEW

**Comment by hjacobs** [comment]
@tuxlife please review :smile:

File: senza/components/redis_cluster.py [hunk 1]
```diff
@@ -21,7 +21,6 @@ def component_redis_cluster(definition, configuration, args, info, force, accoun
             "EngineVersion": configuration.get('EngineVersion', '2.8.19'),
             "CacheParameterGroupName": configuration.get('CacheParameterGroupName', 'default.redis2.8'),
             "NumCacheClusters": number_of_nodes,
-            "CacheNodeType": configuration.get('CacheNodeType', 'cache.t2.small'),
```
Comment (tuxlife) [inline senza/components/redis_cluster.py]: @hjacobs Why did you remove this line?

Comment (hjacobs) [inline senza/components/redis_cluster.py]: the key was duplicate (codacy.com showed this as an issue)

Comment (hjacobs) [inline senza/components/redis_cluster.py]: https://www.codacy.com/app/henning/senza/dashboard

File: senza/cli.py [hunk 2]
```diff
@@ -1203,6 +1206,99 @@ def dump(stack_ref, region, output):
...
+    region = get_region(region)
+    check_credentials(region)
+
+    for asg_name in get_auto_scaling_groups(stack_refs, region):
+        respawn_auto_scaling_group(asg_name, region)
+
+
+@cli.command()
+@click.argument('stack_ref', nargs=-1)
+@click.argument('desired_capacity', type=click.IntRange(0, 100, clamp=True))
```
Comment (tuxlife) [inline senza/cli.py]: Why a hard limit of 100?

File: senza/respawn.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            act.progress()
+            instances_in_service = get_instances_in_service(group, region)
+
+
+def do_respawn_auto_scaling_group(asg_name: str, group: dict, region: str,
+                                  instances_to_terminate: set, instances_ok: set):
+    asg = boto3.client('autoscaling', region)
+    with Action('Suspending scaling processes for {}..'.format(asg_name)):
+        asg.suspend_processes(AutoScalingGroupName=asg_name, ScalingProcesses=SCALING_PROCESSES_TO_SUSPEND)
+    new_min_size = group['MinSize'] + 1
```
Comment (tuxlife) [inline senza/respawn.py]: scale_out only, if the MaxSize/DesiredCapacity 1.

I think, etcd had problems with an additional node, but not with a stopped node.

File: senza/cli.py (hunk 2 above)
Comment (hjacobs) [inline senza/cli.py]: just a sanity check, we can also use another limit or remove it (but at least >= 0)

Comment (tuxlife) [inline senza/cli.py]: :+1:

File: senza/respawn.py (hunk 3 above)
Comment (hjacobs) [inline senza/respawn.py]: the whole respawn stuff is not really supported yet for clustered apps (would need a lot of discussion and testing) --- assuming stateless services right now

Comment (tuxlife) [inline senza/respawn.py]: Make it with a option => inplace-respawn for respawn without scale_out

Comment (hjacobs) [inline senza/respawn.py]: OK, but what would be the behavior for "--inplace" with only 1 instance? This would mean having a downtime (treating "--inplace" and 1 instance specially would be unexpected/inconsistent).

I can live with having an "--inplace" option, but then it means it's really dangerous for "unaware" users.

Comment (tuxlife) [inline senza/respawn.py]: You can Verify the Opton with a Prompt, if the only one instance ;)


Gruß,
Matthias

-- Matthias Kerk | +49 30 2000 88 324 | +49 176 12 758 240 |
matthias.kerk@zalando.de


2015-11-03 22:35 GMT+01:00 Henning Jacobs <notifications@github.com>:
